core/src/main/java/org/apache/calcite/rel/core/Uncollect.java

```diff
@@ -123,14 +126,24 @@ public static RelDataType deriveUncollectRowType(RelNode rel,
     RelDataType inputType = rel.getRowType();
     assert inputType.isStruct() : inputType + " is not a struct";
     final List<RelDataTypeField> fields = inputType.getFieldList();
-    final RelDataTypeFactory.Builder builder =
-        rel.getCluster().getTypeFactory().builder();
+    final RelDataTypeFactory typeFactory = rel.getCluster().getTypeFactory();
+    final RelDataTypeFactory.Builder builder = typeFactory.builder();
     for (RelDataTypeField field : fields) {
       if (field.getType() instanceof MapSqlType) {
         builder.add(SqlUnnestOperator.MAP_KEY_COLUMN_NAME, field.getType().getKeyType());
         builder.add(SqlUnnestOperator.MAP_VALUE_COLUMN_NAME, field.getType().getValueType());
       } else {
         RelDataType ret = field.getType().getComponentType();
+        if (ret == null) {
```

Comment:
Before this change, I assume the assert on line 147 was being hit. What is the scenario where the return is null ?  I am wondering whether it is always legal to do this conversion to a DYNAMIC_STAR .   
Also, should we add a check for `field.getType() instanceof ArraySqlType` similar to the check for MapSqlType on line 129 ?

Reply:
Agreed that DYNAMIC_STAR is reasonable to use here and handle the projection pushdown to uncollect in the future.